internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,162 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/internal/keys"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+var _ storage.OpenFGADatastore = (*cachedDatastore)(nil)
+
+type cachedDatastore struct {
+	storage.OpenFGADatastore
+
+	cache        storage.InMemoryCache[any]
+	maxCacheSize int64
+	ttl          time.Duration
+}
+
+// NewCachedDatastore returns a wrapper over a datastore...
+func NewCachedDatastore(inner storage.OpenFGADatastore, cache storage.InMemoryCache[any], maxSize int64, ttl time.Duration) *cachedDatastore {
+	return &cachedDatastore{
+		OpenFGADatastore: inner,
+		cache:            cache,
+		maxCacheSize:     maxSize,
+		ttl:              ttl,
+	}
+}
+
+func (c *cachedDatastore) ReadUsersetTuples(ctx context.Context, store string, filter storage.ReadUsersetTuplesFilter, options storage.ReadUsersetTuplesOptions) (storage.TupleIterator, error) {
+	iter, err := c.OpenFGADatastore.ReadUsersetTuples(ctx, store, filter, options)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheKey := fmt.Sprintf("%s/%s#%s", store, filter.Object, filter.Relation)
+	// todo(jpadilla): stable sort filter.AllowedUserTypeRestrictions
+	for _, userset := range filter.AllowedUserTypeRestrictions {
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Relation); ok {
+			cacheKey += "/" + userset.GetType() + ":%#" + userset.GetRelation()
+		}
+		if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Wildcard); ok {
+			cacheKey += "/" + userset.GetType() + ":*"
+		}
+	}
```

Comment:
for efficient string building use a [strings.Builder](https://pkg.go.dev/strings#Builder)